File: src/main/scala/werger/adapters/db/DbConfig.scala
package werger.adapters.db

import cats.effect.IO

final case class DbConfig(host: String, port: Int, user: String, password: String, database: String)

object DbConfig:
  private def required(lookup: String => Option[String], name: String): IO[String] =
    IO.fromOption(lookup(name))(new NoSuchElementException(s"Missing required env var: $name"))

  def fromEnv(lookup: String => Option[String]): IO[DbConfig] =
    for
      host <- required(lookup, "DB_HOST")
      port <- IO(lookup("DB_PORT").getOrElse("6543").toInt)
      user <- required(lookup, "DB_USER")
      password <- required(lookup, "DB_PASSWORD")
      database <- required(lookup, "DB_NAME")
    yield DbConfig(host, port, user, password, database)

File: src/main/java/werger/adapters/db/DbConfig.java
package werger.adapters.db;

import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.function.Function;

public record DbConfig(String host, int port, String user, String password, String database) {
    private static String required(Function<String, Optional<String>> lookup, String name) {
        return lookup.apply(name).orElseThrow(() -> new NoSuchElementException("Missing required env var: " + name));
    }

    /** Reads connection settings through {@code lookup} so tests never have to touch the real environment. */
    public static DbConfig fromEnv(Function<String, Optional<String>> lookup) {
        return new DbConfig(
                required(lookup, "DB_HOST"),
                Integer.parseInt(lookup.apply("DB_PORT").orElse("6543")),
                required(lookup, "DB_USER"),
                required(lookup, "DB_PASSWORD"),
                required(lookup, "DB_NAME"));
    }

    /**
     * {@code prepareThreshold=0} stops pgjdbc from ever promoting a statement to a named server-side prepared
     * statement. Supavisor's transaction pooler (port 6543) hands out a backend connection per transaction, not per
     * session, so a named statement prepared on one backend can be missing or collide on the next.
     */
    public String jdbcUrl() {
        return "jdbc:postgresql://" + host + ":" + port + "/" + database + "?prepareThreshold=0";
    }
}
